cirq/protocols/json_serialization.py

```diff
@@ -421,6 +425,19 @@ def _cirq_object_hook(d, resolvers: Sequence[JsonResolver]):
             "Could not resolve type '{}' during deserialization".format(d['cirq_type'])
         )
 
+    if cls == _SerializedKey:
+        read_from_context = getattr(cls, 'read_from_context', lambda x, y: None)
+        return read_from_context(context_map, **d)
+
+    if cls == _SerializedContext:
+        update_context = getattr(cls, 'update_context', lambda x, y: None)
+        update_context(context_map, **d)
+        return None
+
+    if cls == _ContextualSerialization:
+        deserialize_with_context = getattr(cls, 'deserialize_with_context', lambda x: None)
+        return deserialize_with_context(**d)
+
```

Comment:
What if we actually instead of making `_from_json_dict_` throw an error, we would just make it do `update_context`, `deserialize_with_context`, etc.? Why not?

Reply:
If we use `_from_json_dict` then it makes sense to keep these classes in the resolvercache..

Reply:
My personal sense is that implementing this behavior in `_from_json_dict_` violates an implicit contract of the method - specifically, that the following code returns the original object:
```
def json_roundtrip(obj):
    cls = type(obj)
    return cls._from_json_dict_(object._json_dict_())
```
While we probably could get away with this since these are purely internal types, I prefer making the distinction explicit so as not to confuse future work in this area.

Reply:
Sounds good - I see your point. Let's leave them then for clarity.